Block validation: Ensure transactions are sorted

Geod24 opened the pull request:
So that we get a more stable block creation *and* can leverage this for block propagation.

Geod24 committed:
Update BitBlob to v1.1.4

Geod24 committed:
Block validation: Ensure transactions are sorted

AndrejMitrovic commented:
LGTM. The docker failure is probably because the `getChainedTransaction` function is duplicated for the integration tests. So there are missing `sort` calls somewhere in those tests.

AndrejMitrovic commented:
> LGTM. The docker failure is probably because the getChainedTransaction function is duplicated for the integration tests. So there are missing sort calls somewhere in those tests.

Well I expect this to become an error later, but for now it seems something else is wrong https://circleci.com/gh/bpfkorea/agora/29?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Geod24 committed:
CircleCI: Checkout merge commit instead of PR HEAD

Geod24 committed:
CircleCI: Install ssh so that we use real git

Without ssh, CircleCI uses their native client,
which, contrary to what the documentation states,
attempt to checkout submodules.
It seems to do some wrong caching,
leading to failing PRs when we attempt to upgrage submodules.

Geod24 committed:
CircleCI: Sync submodule before updating them

Geod24 commented:
Slapped the CI into submission.

Geod24 commented on source/agora/consensus/data/Block.d:
```diff
@@ -374,6 +374,7 @@ public Block makeNewBlock (const ref Block prev_block, Transaction[] txs) @safe
     block.header.prev_block = prev_block.header.hashFull();
     block.header.height = prev_block.header.height + 1;
     block.txs ~= txs;
+    block.txs.sort!((a, b) => hashFull(a) < hashFull(b));
```
I think I can remove the predicate now

Geod24 replied:
Done

Geod24 commented:
Green!
